### DISCUSSION TITLE
Possible ways to extend/reuse existing machines?

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#2378] [Q&A] [answered]
Hello everyone.

I have a use case where I need to create a new machine by extending an existing machine definition with extra events/actions, parallel/nested states, and context fields.

What are the possible ways to do so?

I could think of merging definitions. Any better ways/caveats to look for?

**Comment by davidkpiano** [comment, marked as answer]
> I could think of merging definitions. Any better ways/caveats to look for?

I think this would be a good idea, and I know that others have done this before. We'd have to test this out before recommending it as an official pattern, and creating helpers for it 🤔

**Reply by geekox86** [reply]
So I tried to merge multiple machine configuration and option objects to extend my new machine from existing ones.

The only problem that I had so far is conflicting names (fields, states, events, actions, guards, ...etc.).

I tried to solve by using symbols but I came to know that xstate does not support them instead of strings.

I wanted to have multiple lookup dictionaries of field, state, event, ... etc. symbols. But when I use the [fields.someSymbol] I immediately loose the static typing of TypeScript.

So, any thoughts on how to resolve this problem or improve it?

Using string prefixes makes everything long and hard to read.

**Reply by davidkpiano** [reply]
Can you share a code example of using string prefixes?
